python/ray/ml/checkpoint.py

```diff
@@ -368,8 +398,24 @@ def as_directory(self) -> Iterator[str]:
             yield self._local_path
         else:
             temp_dir = self.to_directory()
+            del_lock_path = _get_del_lock_path(temp_dir)
```

Comment:
This seems sketchy. How about we just skip file locking for as_directory() and don't re-use directories?

We should also document to_directory() to say that you shouldn't delete the directory unless you specify a specific path, since it may be shared.

Reply:
The predictors called by BatchPredictor use `as_directory`. We need to either not delete the directory after a worker stops using it (leading to wasted disk space) or ensure that only the last worker deletes it. Otherwise we can have a situation where the directory is deleted while it is being read by another worker, or it would force the files to be copied again, negating the performance gain.